sdks/python/apache_beam/ml/inference/tensorrt_inference.py

```diff
@@ -0,0 +1,281 @@
+          'name': name,
+          'dtype': np.dtype(trt.nptype(dtype)),
+          'shape': list(shape),
+          'allocation': allocation,
+          'size': size
+      }
+      self.gpu_allocations.append(allocation)
+      if self.engine.binding_is_input(i):
+        self.inputs.append(binding)
+      else:
+        self.outputs.append(binding)
+
+    assert self.context
+    assert len(self.inputs) > 0
+    assert len(self.outputs) > 0
+    assert len(self.gpu_allocations) > 0
+
+    for output in self.outputs:
+      self.cpu_allocations.append(np.zeros(output['shape'], output['dtype']))
+    # Create CUDA Stream.
+    self.stream = ASSERT_DRV(cuda.cuStreamCreate(0))
+
+  def get_engine_attrs(self):
+    """Returns TensorRT engine attributes."""
+    return (
+        self.engine,
+        self.context,
+        self.inputs,
+        self.outputs,
+        self.gpu_allocations,
+        self.cpu_allocations,
+        self.stream)
+
+
+class TensorRTEngineHandlerNumPy(ModelHandler[np.ndarray,
+                                              PredictionResult,
+                                              TensorRTEngine]):
+  def __init__(self, min_batch_size: int, max_batch_size: int, **kwargs):
+    """Implementation of the ModelHandler interface for TensorRT.
+
+    Example Usage::
+
+      pcoll | RunInference(
+          TensorRTEngineHandlerNumPy(
+            min_batch_size=1,
+            max_batch_size=1,
+            engine_path="my_uri"))
+
+    Args:
+      min_batch_size: minimum accepted batch size.
+      max_batch_size: maximum accepted batch size.
+      kwargs: Additional arguments like 'engine_path' and 'onnx_path' are
+        currently supported.
+
+    See https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/
+    for details
+    """
+    self.min_batch_size = min_batch_size
+    self.max_batch_size = max_batch_size
+    if 'engine_path' in kwargs:
+      self.engine_path = kwargs.get('engine_path')
+    elif 'onnx_path' in kwargs:
+      self.onnx_path = kwargs.get('onnx_path')
+
+    trt.init_libnvinfer_plugins(TRT_LOGGER, namespace="")
+
+  def batch_elements_kwargs(self):
+    """Sets min_batch_size and max_batch_size of a TensorRT engine."""
+    return {
+        'min_batch_size': self.min_batch_size,
+        'max_batch_size': self.max_batch_size
+    }
+
+  def load_model(self) -> TensorRTEngine:
+    """Loads and initializes a TensorRT engine for processing."""
+    engine = _load_engine(self.engine_path)
+    return TensorRTEngine(engine)
+
+  def load_onnx(self) -> Tuple[trt.INetworkDefinition, trt.Builder]:
+    """Loads and parses an onnx model for processing."""
+    return _load_onnx(self.onnx_path)
+
+  def build_engine(
+      self, network: trt.INetworkDefinition,
+      builder: trt.Builder) -> TensorRTEngine:
+    """Build an engine according to parsed/created network."""
+    engine = _build_engine(network, builder)
+    return TensorRTEngine(engine)
+
+  def run_inference(
```

Comment:
Is this implementation thread-safe?
Beam could potentially have multiple concurrent threads within a process. Are there any concerns with that from hardware or software standpoint?

Reply:
Looks thread-safe. It would be good to share the engine between threads so you don't end up duplicating the model weights (maybe that can be future work). Each thread will still need a separate execution context.

Reply:
Ok, leave it for the future. Big thanks @pranavm-nvidia, any tips where I can read more about how to properly implement this?

Reply:
Makes sense. Thank you @pranavm-nvidia!

Reply:
So, I think the return result of the load_model (i.e. the engine) will be shared between threads: https://github.com/apache/beam/pull/22131/files#diff-145530e8e39084aed7792a4d3b8104258ea1635db466ab48a58f168bce10dde3R204

OTOH, it looks like we only create the execution context once, when we initialize the engine: https://github.com/apache/beam/pull/22131/files#diff-145530e8e39084aed7792a4d3b8104258ea1635db466ab48a58f168bce10dde3R114

in this case, the context and the engine will both be shared across threads.

Reply:
we could introduce a context factory that will have a thread-level cache.

Reply:
@tvalentyn Any examples you can point me to implement something like this. I want to make sure I follow your specific vision of this.

Reply:
I was thinking something along the lines of having a dictionary[Thread ID -> execution_context], and create the context if it has not been created by this thread yet, otherwise fetch it from the dictionary.

Is it expensive in any sense to create execution contexts? As in, would it make sense to try to recycle the context between multiple non-concurrent threads (i.e. reuse the context after a thread that created it has finished)?

Reply:
I think @pranavm-nvidia can give a better estimate on how expensive this might me.

Reply:
Yes, it can be expensive, so it would make sense to reuse. How frequently do you expect new threads to be spawned?

Reply:
TLDR: we reuse threads, but there can be some churn.

Here's the relevant code:
https://github.com/apache/beam/blob/9154f8b3cf01fce067cea7ad7522e1db3154ff89/sdks/python/apache_beam/runners/worker/sdk_worker.py#L352-L361
https://github.com/apache/beam/blob/48b534939468cc8e6dd403f11689a4433ca2e5a4/sdks/python/apache_beam/utils/thread_pool_executor.py#L82-L107

The treads are managed by a ThreadPool, which is backed by a LIFO queue. Inactive threads can be removed from the queue if their number exceed a threshold. 
W.r.t. our usecase, we could potentially consider modifying the underlying collection to be a deque, so that inactive threads are removed from one end, and use LIFO so there is a higher chance of reusing the same threads. 

In case of Batch jobs, I think the thread queue is low, in case of streaming jobs, I am not sure, likely low after all threads have spawned unless there is some bursty behavior. I can ask around. Could you also elaborate a bit now on the costs for creating the context?

Reply:
> The treads are managed by a ThreadPool, which is backed by a LIFO queue.

Correct me if I overlooked something, but aren't we [instantiating](https://docs.python.org/3/library/queue.html#queue.Queue) a FIFO queue here?
https://github.com/apache/beam/blob/48b534939468cc8e6dd403f11689a4433ca2e5a4/sdks/python/apache_beam/utils/thread_pool_executor.py#L76

Reply:
Yes, we have a FIFO currently, that was a typo.

Reply:
If precious memory is allocated that is already a good motivation to reuse. That means, we'd have to also think how garbage collection would work so that creating these contexts wouldn't cause OOMs in a long running streaming job.

Reply:
Another possibility is to introduce a lock so that we the same context, but run in one thread at a time. As we know due to GIL there is a limit to thread-parallelism in Python within a single process. So perhaps it wouldn't have significant impact on actual performance? I suppose this needs some experimental evaluation to confirm. Or is that obviously a bad idea?

Reply:
All the long running functions in `IExecutionContext` (`execute_async_v2`, `execute_v2`, etc.) release the GIL while they're running, so I think there could be a noticeable performance impact. 

One way to mitigate the possibility of OOMs is to use [`create_execution_context_without_device_memory`](https://docs.nvidia.com/deeplearning/tensorrt/api/python_api/infer/Core/Engine.html#tensorrt.ICudaEngine.create_execution_context_without_device_memory) which will let you bring your own memory.

Reply:
> which will let you bring your own memory.

What would we need to do to manage memory in this case?

Reply:
The only requirement is that you set `context.device_memory` to point to a region of GPU memory at least as large as `engine.device_memory_size` (which is the size in bytes). You'd still need to make sure you're not using the same memory from different threads at the same time, but you would at least have a greater degree of control over the allocations.

Reply:
Got it, I think I understand. Is engine.device_memory_size defined by the tensorrt model?

Reply:
Yes, `device_memory_size` is tied to the model/engine.

Reply:
Another thing we have to consider is if users are using multiple GPUs. For Dataflow jobs, one would modify the count like so:
```
python -m apache_beam.examples.inference.tensorrt_object_detection \
--experiment="worker_accelerator=type:nvidia-tesla-t4;count:1;install-nvidia-driver" \
...
```
or
```
python -m apache_beam.examples.inference.tensorrt_object_detection \
--experiment="worker_accelerator=type:nvidia-tesla-t4;count:2;install-nvidia-driver" \
...
```
I'm wondering the best way to manage this. For example, an easy scenario is if we had 4 threads, and 1 GPU. But it would be a bit more complicated if there were 4 threads and 2 GPUs, or 4 threads and 4 GPUs.

A user in this forum [discussion](https://github.com/NVIDIA/TensorRT/issues/219#issuecomment-559249117) was asking how to properly deal with multithreading with multiple GPUs. The solution was to make a custom class inheriting from Thread, where each thread has the correct mapping of engine and context for the correct GPU.

Reply:
Using multiple GPUs efficiently may not be straightforward with Beam Python. I think we'd need to understand how common users would need something like this - users would need to utilize a single gpu to the fullest extent first, and it should be a condition where adding another GPU outweighs starting another worker with a single GPU.

Reply:
@yeandy When you are talking about multi-GPU. What are you envisioning as a workflow? Do you want to split TRT model between multiple-GPUs? Or if it is a multi-GPU machine you want to selected a specific one and use it for inference?

Reply:
I was originally thinking of the latter. But now that I think about it again, I'm not sure how much utility, if any, we'd get out of trying to work with multiple GPUs.. We're already using Beam's `Shared` class to share both context and the engine across threads. We'd want to first resolve the thread-safety concern with 1 GPU. 

For the former, I reckon that probably require a significant rewrite of the logic, right? It would definitely be out of scope of this PR.

Reply:
@yeandy TRT does not currently support multi-GPU so we don't need to worry about that. Meaning you cannot take one big model and split it across multiple GPUs (model parallelism). 

In terms of multi-GPU machine and selecting specific GPU, we can just use [env variable](https://stackoverflow.com/questions/39649102/how-do-i-select-which-gpu-to-run-a-job-on) or whatever the underlying mechanism that Beam has. I guess it's more of a question how is Beam architected with regards to this and if you see a solution and need our advice, please ask. Maybe if solution is not clear, you can outline the use case and how we should get to solution abstractly. After that we can do detailed back and forth on how we can do this.

Reply:
> We'd want to first resolve the thread-safety concern with 1 GPU.

We need to address the thread safety concern per feedback in this thread. At minimum, let's add a lock and avoid sharing the context. 

I  discussed this a bit with lcwik. In short term, we should implement a shared object cache for managing shared execution contexts. We should avoid the affinity between thread IDs and contexts, and instead make it possible to transfer shared contexts between threads.

We should be able to rely on `start_bundle` and `finish_bundle` to claim and release the contexts from the shared pool.  

Open question is GPU memory management:
- How many contexts do we create and how much GPU memory should we allocate per each context. Can we know at runtime what's the max # of context we can create for a particular engine? What are the tradeoffs between creating more contexts with less memory per contexts vs less contexts? Will pipeline degrade gracefully when more memory is needed or something will crash? If we solve memory management, adding the shared pool should not be too complex. We can refer to `UnboundedThreadPoolExecutor` implementation for inspiration.

- In the future, we can consider sharing the object cache w/ contexts between multiple SDK processes (not just between threads) via `multiprocessing`. This may be worthwhile with A100 accelerators that have more GPU memory that can be shared across multiple processes, and this could allow users benefiting from MPS.  We'd have to share contexts between processes - are these objects large in size? If it is just metadata, the overhead may not be significant.  This is something we should also consider for Tensorflow runinference, since Tensorflow does not work well when using the same GPU from multiple processes and manual GPU memory management is needed, so Beam RunInference could help make it easier.  In the medium term, we may need to recommend limiting pipelines to use 1 process only.

Summary of questions asked above:
- How many contexts can we create and how do you recommend to manage memory b/w them?
- Are context objects small?
- Does TensorRT support concurrent execution on multiple processes on the same GPU? I believe, I asked this and answer was `yes`, just, double checking...  If yes, how GPU memory management is solved?

Reply:
> we should implement a shared object cache

We can take over this work and I think I know someone who may be interested to help here, however I'd like to understand the limits of parallelization we can plan for. Do you have insights to questions above @pranavm-nvidia ?

Reply:
We will have an internal sync this week to try to answer as much as possible memory related questions. I think it is not scheduled yet, but should happen this week. @tvalentyn My apologies, I should've kept you updated

Reply:
Like @azhurkevich said, we'll have more details after our internal sync, but I can provide a little background information until then.

The context mainly allocates three things:
1. Persistent host/device memory: This is for things like look-up tables to speed up computation.
2. Workspace/scratch memory: This is memory that is reused across layers and used during computation - e.g. if we compute Convolution as a GEMM, the generated matrix would be written to scratch memory.
3. Activation memory: This is to store intermediate tensors. TensorRT tries to reuse this across layers as much as possible, but the degree of reusability depends on the network architecture (e.g. parallel branches would require separate memory).

You can determine the total size of (2) and (3) by checking `engine.device_memory_size` and optionally manage both of these yourself by using `engine.create_execution_context_without_device_memory()` and setting `context.device_memory` to point to your own memory buffer. Otherwise, TensorRT will allocate them on a per-context basis. 

Finally, to answer your questions:

> How many contexts can we create and how do you recommend to manage memory b/w them?

There's some nuance here, but effectively you're only limited by system resources. Regarding memory management, it depends on the use-case. Assuming you have N contexts, then:
- If you plan to use all N concurrently, then I'd just let TensorRT manage the per-context memory
- If you plan to use only some fraction, K, of them concurrently, then you can allocate enough memory for only K contexts and reuse that. The thing to be careful of is that contexts executing concurrently should not use the same memory.

> Are context objects small?

Generally no.
    
> Does TensorRT support concurrent execution on multiple processes on the same GPU? I believe, I asked this and answer was yes, just, double checking... If yes, how GPU memory management is solved?

Yes it does, but the CUDA context switching overhead might make this impractical (I'm not totally sure though, will need to double check this). For GPU memory management, I think it might be possible to use `nvshmem` to share memory across processes, but I don't have any experience with that.

Reply:
Great, thank you for helping to investigate this!

> If you plan to use all N concurrently, then I'd just let TensorRT manage the per-context memory

The assumption is that contexts will be used concurrently. Is the per-context allocation static? I think we will have to choose between: 
- creating contexts on demand as new threads spun up, possibly with a maximum limit for total # of contexts
- pre-creating a limited number of contexts and block threads until shared contexts become available, or just limit thread parallelism to the number of context that we can afford for a particular engine.

Reply:
For models with static shapes, yes, the memory allocation would be static. When dynamic shapes are involved, there may be reallocations when the input shapes change. 

This:
> just limit thread parallelism to the number of contexts that we can afford for a particular engine

is required for both options. 
So then it's just a question of whether to create contexts up front or on-demand. I think the answer to that depends on what the typical usage scenario is.

Reply:
Per meeting offline, let's try to go with having one context only for now and add use a lock where necessary to guard access to it. In this case, if there are multiple threads, they will have to wait.
Later, we will conduct experiments evaluating GPU utilization and deciding whether having multiple contexts would be beneficial. @azhurkevich would you be comfortable making this change?

Reply:
Not sure what kind of change is required considering the way it is built rn is that it will be one context. With regards to locking mechanism, that one is definitely not in place, I wonder if it'll happen automatically. @pranavm-nvidia any ideas about Valentin's request?

Reply:
Sounds like `threading.Lock` should work here? Unless there is a separate API provided by Beam which would be preferred?

Reply:
@tvalentyn any thoughts?

Reply:
We could use threading.RLock, so that the same thread can also aquire it if it somehow becomes necessary, I haven't look very close into where to place this lock.

Reply:
@tvalentyn Do we have this mechanism that you are mentioning in PyTorch example? It would be easier for me to understand what you would like to see if there is a reference.

Reply:
no. I took a closer look, see suggestions. I haven't tested it.